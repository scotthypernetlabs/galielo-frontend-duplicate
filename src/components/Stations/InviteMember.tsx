import React from 'react';
import { connect } from 'react-redux';
import { IStore } from '../../business/objects/store';
import { Dispatch } from 'redux';
import { User, UserFilterOptions } from '../../business/objects/user';
import { Station } from '../../business/objects/station';
import { UserIconNew } from '../svgs/UserIconNew';
import { MyContext } from '../../MyContext';
import { context } from '../../context';
import { IReceiveSearchedUsers, receiveSearchedUsers } from '../../actions/userActions';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import {closeModal, ICloseModal} from "../../actions/modalActions";

interface MatchParams {
  id: string;
}

interface Props extends RouteComponentProps<MatchParams> {
  predictions: User[];
  station: Station;
  currentUser: User;
  receiveSearchedUsers: (user_list: User[]) => IReceiveSearchedUsers;
  closeModal: () => ICloseModal;
}

type State = {
  searchInput: string;
}

const updateState = <T extends string>(key: keyof State, value: T) => (
  prevState: State
): State => ({
  ...prevState,
  [key]: value
})

class InviteMembers extends React.Component<Props, State> {
  context!: MyContext;
  constructor(props: Props){
    super(props);
    this.state = {
      searchInput: ''
    }
    this.removeUserFromStation = this.removeUserFromStation.bind(this);
    this.inviteUserToCurrentStation = this.inviteUserToCurrentStation.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.displayPredictions = this.displayPredictions.bind(this);
  }
  public removeUserFromStation(user_id: string){
    return(e:any) => {
      this.context.stationService.expelUser(this.props.station.id, user_id);
    }
  }
  public inviteUserToCurrentStation(user_id: string){
    return(e:any) => {
      this.context.stationService.inviteUsersToStation(this.props.station.id, [user_id]);
    }
  }
  public handleChange(type:keyof State){
    return(e:React.ChangeEvent<HTMLInputElement>) => {
      let value = e.target.value;
      this.setState(updateState(type, value));
      if(value.length > 1){
        this.context.userService.searchByUsername(new UserFilterOptions(null, value));
      }else{
        this.props.receiveSearchedUsers([]);
      }
    }
  }
  public displayPredictions(){
    const { predictions, station } = this.props;
    return(
      <div className="predictions">
      {
        predictions.map((prediction, idx) => {
          let inStationAlready = station.members.indexOf(prediction.user_id) >= 0;
          let alreadyInvited = false;
          if(station.invited_list.indexOf(prediction.user_id) >= 0){
            alreadyInvited = true;
          }
          if(prediction.user_id === this.props.currentUser.user_id){
            return;
          }
          return(
            <div key={`${prediction.user_id}${idx}`} className="prediction">
              {UserIconNew('OFFLINE', 30)}
              <div className="prediction-name">{prediction.username}</div>
              <div className="member-invite-button">
              {
                inStationAlready &&
                station.admins.indexOf(this.props.currentUser.user_id) >= 0 &&
                <button className="secondary-btn" onClick={this.removeUserFromStation(prediction.user_id)}>
                  Remove
                </button>
              }
              {
                  alreadyInvited && <div>Invited</div>
              }
              {
                !inStationAlready && !alreadyInvited &&
                <button className="secondary-btn" onClick={this.inviteUserToCurrentStation(prediction.user_id)}>
                  Invite
                </button>
              }
              </div>
            </div>
          )
        })
      }
      </div>
    )
  }
  render(){
    return(
      <div className="modal-style" onClick={(e:any) => e.stopPropagation()}>
        <div className="invite-members">
          <div className="group-machine-modal-title">
            <span>Add Users</span>
            <div onClick={this.props.closeModal}><i className="fal fa-times"/></div>
          </div>
          <button className="user-search">
            <div className="user-search-inner">
              <input
                className="user-search-input"
                type="text"
                value={this.state.searchInput}
                onChange={this.handleChange("searchInput")}
                placeholder={"Search by Email"}
                />
            </div>
          </button>
          {this.displayPredictions()}
        </div>
      </div>
    )
  }
}

InviteMembers.contextType = context;

type InjectedProps = {
  match: any,
  history: any
}

const mapStateToProps = (store: IStore, ownProps: InjectedProps) => {
  if(!store.stations.stations[ownProps.match.params.id]) {
    return({
      predictions: store.users.searchedUsers,
      currentUser: store.users.currentUser,
      stations: store.stations.stations,
      station: {}
    })
  }

  return({
    predictions: store.users.searchedUsers,
    currentUser: store.users.currentUser,
    stations: store.stations.stations,
    station: store.stations.stations[ownProps.match.params.id]
  })
};

const mapDispatchToProps = (dispatch:Dispatch) => ({
  receiveSearchedUsers: (user_list: User[]) => dispatch(receiveSearchedUsers(user_list)),
  closeModal: () => dispatch(closeModal())
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(InviteMembers));
