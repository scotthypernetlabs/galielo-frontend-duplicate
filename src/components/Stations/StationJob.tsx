import React from 'react';
import { connect } from 'react-redux';

type Props = {
  job: any;
}
type State = {
  counter: number;
  showDropdown: boolean;
  showContent: boolean;
  timer: boolean;
  currentlyDownloading: boolean;
}

class StationJob extends React.Component<Props, State>{
  clockTimer: any;
  constructor(props: Props){
    super(props);
    this.state = {
      counter: 0,
      showDropdown: false,
      showContent: false,
      timer: false,
      currentlyDownloading: false
    }
  }
  public componentDidMount() {
    this.clockTimer = setInterval(() => {
      this.setState(prevState => { return { counter: prevState.counter + 1, timer: true } })
    }, 1000);
    let time_elapsed = this.props.job.run_time;
    // let storedTime;
    // storedTime = this.props.stationJobTimer[this.props.job.id];
    // if(storedTime && storedTime > time_elapsed){
      // time_elapsed = storedTime;
    // }
    this.setState({
      counter: time_elapsed
    })
  }
  public componentWillUnmount(){
    clearInterval(this.clockTimer);
    // this.props.stationJobTimer(this.props.job.id, this.state.counter);
  }
  public startTimer(){
    if(this.state.timer){
      return;
    }
    this.setState({
      counter: this.props.job.run_time
    })
  }
  public stopTimer(){
    clearInterval(this.clockTimer);
    this.setState({ timer: false, counter: this.props.job.run_time })
  }
  public parseTime(seconds_elapsed:number, timeString:string = "", shortDuration:boolean = true):string{
    if(seconds_elapsed >= 3600){
      let hours:number = Math.floor(seconds_elapsed / 3600);
      let hours_string: string = hours.toString();
      if(hours < 10){
        hours_string = '0' + hours;
      }
      return this.parseTime(seconds_elapsed - (hours * 3600),`${hours_string}`, false);
    }else{
      let minutes = 0;
      let seconds = seconds_elapsed;
      let minutes_string = minutes.toString();
      let seconds_string = seconds.toString();
      if(seconds_elapsed >= 60){
        minutes = Math.floor(seconds_elapsed / 60);
        minutes_string = minutes.toString();
        seconds = seconds_elapsed - (minutes * 60);
        seconds_string = seconds.toString();
      }
      if(seconds < 10){
        seconds_string = '0' + seconds;
      }
      if(minutes < 10){
        minutes_string = '0' + minutes;
      }
      if(shortDuration){
        return `00:${minutes_string}:${seconds_string}`;
      }else{
        return timeString + `:${minutes_string}:${seconds_string}`;
      }
    }
  }
  render(){
    const { job } = this.props;
    let landingZone = job.landing_zone;
    let launchPad = job.launch_pad;
    let elapsedTime;
    let uploadTime = Math.floor(job.upload_time);
    let currentTime = Math.floor(new Date().getTime() / 1000);
    if(job.status === 'Running'){
      elapsedTime = currentTime - uploadTime;
    }else{
      elapsedTime = job.run_time;
    }
    if(this.state.counter > elapsedTime){
      elapsedTime = this.state.counter;
    }
    return(
      <>
        <div className="log-column">
          <div className="station-job-info">
                <div className="submitted-by">{landingZone}</div>
                <div className="submitted-by">{launchPad}</div>
                <div className="project-name">{job.name}</div>
                <div className="job-time-taken">{this.parseTime(elapsedTime)}</div>
          </div>
        </div>
      </>
    )
  }
}

export default StationJob;
