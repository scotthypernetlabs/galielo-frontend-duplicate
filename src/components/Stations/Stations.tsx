import React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { IStore } from '../../business/objects/store';
import { Link } from 'react-router-dom';
import { Station } from '../../business/objects/station';
import { Dictionary } from '../../business/objects/dictionary';
import { openModal, IOpenModal } from '../../actions/modalActions';
import { MyContext } from '../../MyContext';
import { context } from '../../context';
import {Box, Button, Grid, Typography} from "@material-ui/core";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faChalkboard, faDatabase, faUser} from "@fortawesome/free-solid-svg-icons";

type Props = {
  stations: Dictionary<Station>;
  openCreateStation: () => IOpenModal;
}

type State = {
}

class Stations extends React.Component<Props, State> {
  context!: MyContext;
  constructor(props: Props){
    super(props);
  }
  componentDidMount(){
    this.context.stationService.refreshStations();
  }
  render(){
    if(!this.props.stations){
      return(
        <>
        </>
      )
    }
    return(
      <div className="stations-container">
        <Grid
          container={true}
          justify="space-between"
          alignItems="baseline"
        >
          <Grid item={true}>
            <Typography
              variant="h3"
              style={{fontWeight: 500}}
            >
              Stations
            </Typography>
          </Grid>
          <Grid item={true}>
            <Button
              variant="contained"
              color="primary"
              onClick={this.props.openCreateStation}>
              Add Station
            </Button>
          </Grid>
        </Grid>
        <Grid container={true}>
          {
          Object.keys(this.props.stations).map( (station:any, idx:number) => {
            station = this.props.stations[station];
            if(!station.machines || !station.members || !Object.keys(station.volumes)){
              return (
                <React.Fragment key={idx}>
                </React.Fragment>
              )
            }
            return(
              <Grid container xs={1} sm={2}>
                <Link to={`/stations/${station.id}`} key={station.id} style={{width:"100%"}}>
                  <Box
                    border={1}
                    borderColor="#cccccc"
                    p={3}
                    m={1}
                    minWidth={150}
                    width="85%"
                    height={120}
                    bgcolor="white"
                  >
                    <Grid container>
                      <Grid item={true} xs={12}>
                        <Typography gutterBottom={true} variant="h3" color="primary">{station.name}</Typography>
                      </Grid>
                      <Grid item={true} xs={4}>
                        <FontAwesomeIcon icon={faChalkboard} style={{color: "black", float: 'left', marginRight: 5}}/>
                        <Typography variant="h5">{station.machines.length}</Typography>
                      </Grid>
                      <Grid item={true} xs={4}>
                        <FontAwesomeIcon icon={faUser} style={{color: "black", float: 'left', marginRight: 5}}/>
                        <Typography variant="h5">{station.members.length}</Typography>
                      </Grid>
                      <Grid item={true} xs={4}>
                        <FontAwesomeIcon icon={faDatabase} style={{color: "black", float: 'left', marginRight: 5}}/>
                        <Typography variant="h5">{Object.keys(station.volumes).length}</Typography>
                      </Grid>
                    </Grid>
                  </Box>
                </Link>
              </Grid>
            )
          })
        }
        </Grid>
      </div>
    )
  }
}

Stations.contextType = context;

const mapStateToProps = (state: IStore) => ({
  stations: state.stations.stations
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  openCreateStation: () => dispatch(openModal('Create Station'))
});

export default connect(mapStateToProps, mapDispatchToProps)(Stations);
