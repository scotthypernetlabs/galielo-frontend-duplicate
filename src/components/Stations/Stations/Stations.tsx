import { Dictionary } from "../../../business/objects/dictionary";
import { Dispatch } from "redux";
import { IOpenModal, openModal } from "../../../actions/modalActions";
import {
  IReceiveSelectedStation,
  receiveSelectedStation
} from "../../../actions/stationActions";
import { IStore } from "../../../business/objects/store";
import { MyContext } from "../../../MyContext";
import { RouteComponentProps } from "react-router-dom";
import { Station } from "../../../business/objects/station";
import { User } from "../../../business/objects/user";
import { connect } from "react-redux";
import { context } from "../../../context";
import React from "react";
import WelcomeView from "./WelcomeView";
import StationsView from "./StationsView"
import { Card, Box } from "@material-ui/core";
const fileUploadTextDefault = "Browse or drop directory";


interface Props extends RouteComponentProps<any> {
  slice?: boolean;
  numberOfStations?: number;
  stations: Dictionary<Station>;
  currentUser: User;
  openCreateStation: () => IOpenModal;
  receiveSelectedStation: (station: Station) => IReceiveSelectedStation;
}

type State = {
  dragOver: boolean;
  disabled: boolean;
  fileUploadText: string;
  fileUpload: boolean;
};

class Stations extends React.Component<Props, State> {
  context!: MyContext;
  constructor(props: Props) {
    super(props);
    this.state = {
      dragOver: false,
      disabled: false,
      fileUploadText: fileUploadTextDefault,
      fileUpload: false
    };
  }

  render() {
    if (!this.props.stations) {
      return <></>;
    }

    const { stations, history, currentUser, openCreateStation, numberOfStations } = this.props;

    return (
      <div className="stations-container">
        
        <Card>
        <Box p = {3}>
        {Object.keys(this.props.stations).length > 0 ? (
          <StationsView
            slice = {this.props.slice} 
            numberOfStations = {this.props.numberOfStations}
            openCreateStation = {openCreateStation}
            history = {history}
            stations = {stations}
            currentUser = {currentUser}
          />
        ) : (
          <WelcomeView openCreateStation={openCreateStation} />
        )}
          </Box>
        </Card>
     
      </div>
    );
  }
}

Stations.contextType = context;

const mapStateToProps = (state: IStore) => ({
  stations: state.stations.stations,
  currentUser: state.users.currentUser
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  openCreateStation: () => dispatch(openModal("Create Station")),
  receiveSelectedStation: (station: Station) =>
    dispatch(receiveSelectedStation(station))
});

export default connect(mapStateToProps, mapDispatchToProps)(Stations);
