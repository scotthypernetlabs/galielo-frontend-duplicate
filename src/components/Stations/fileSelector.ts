import { PackagedFile } from "../../business/objects/packagedFile";

const DEFAULT_FILES_TO_IGNORE: string[] = [
  // '.DS_Store', // OSX indexing file
  // 'Thumbs.db'  // Windows indexing file
];

function shouldIgnoreFile(file: PackagedFile) {
  return DEFAULT_FILES_TO_IGNORE.indexOf(file.name) >= 0;
}

/**
 * This is the most horrible function I have ever, ever come across.
 */
function traverseDirectory(
  entry: IFileSystemDirectoryEntry
): Promise<IFileSystemFileEntry[]> {
  const reader = entry.createReader();
  // Resolved when the entire directory is traversed
  return new Promise(resolveDirectory => {
    const iterationAttempts: Promise<IFileSystemFileEntry[]>[] = [];
    const errorHandler = () => {};

    function readEntries() {
      // According to the FileSystem API spec, readEntries() must be called until
      // it calls the callback with an empty array.
      reader.readEntries((batchEntries: IFileSystemEntry[]) => {
        if (!batchEntries.length) {
          // Done iterating this particular directory
          // Condense the results into a single array
          resolveDirectory(
            Promise.all(iterationAttempts).then(
              (fileSystemFileEntries: IFileSystemFileEntry[][]) => {
                return condense(fileSystemFileEntries);
              }
            )
          );
        } else {
          // Add a list of promises for each directory entry.  If the entry is itself
          // a directory, then that promise won't resolve until it is fully traversed.
          const iterationAttemptPromises = new Array<
            Promise<IFileSystemFileEntry[]>
          >();
          for (const batchEntry of batchEntries) {
            if (batchEntry.isDirectory) {
              iterationAttemptPromises.push(
                traverseDirectory(batchEntry as IFileSystemDirectoryEntry)
              );
            } else {
              iterationAttemptPromises.push(
                Promise.resolve([batchEntry as IFileSystemFileEntry])
              );
            }
          }

          // Promise.all() resolves to an array of each result of each promise.
          // We want to just condense them into a single array
          iterationAttempts.push(
            Promise.all(iterationAttemptPromises).then(
              (iterationEntries: IFileSystemFileEntry[][]) => {
                return condense(iterationEntries);
              }
            )
          );

          // Try calling readEntries() again for the same dir, according to spec
          readEntries();
        }
      }, errorHandler);
    }
    // initial call to recursive entry reader function
    readEntries();
  });
}

function condense<T>(arrayOfArrays: T[][]): T[] {
  let retArray = new Array<T>();
  for (const arr of arrayOfArrays) {
    retArray = retArray.concat(arr);
  }
  return retArray;
}

function getFile(entry: IFileSystemFileEntry): Promise<PackagedFile> {
  return new Promise(resolve => {
    entry.file((file: File) => {
      resolve(new PackagedFile(file, entry));
    });
  });
}

function handleFilePromises(
  promises: Promise<PackagedFile>[],
  fileList: PackagedFile[]
): Promise<PackagedFile[]> {
  return Promise.all(promises).then((files: PackagedFile[]) => {
    for (const file of files) {
      if (!shouldIgnoreFile(file)) {
        fileList.push(file);
      }
    }
    return fileList;
  });
}

function getDataTransferFiles(
  dataTransfer: DataTransfer
): Promise<PackagedFile[]> {
  const dataTransferFiles: PackagedFile[] = [];
  const folderPromises: Promise<IFileSystemFileEntry[]>[] = [];
  const filePromises: Promise<PackagedFile>[] = [];

  [].slice.call(dataTransfer.items).forEach((listItem: DataTransferItem) => {
    if (typeof listItem.webkitGetAsEntry === "function") {
      const entry: IFileSystemEntry = listItem.webkitGetAsEntry();

      if (entry) {
        if (entry.isDirectory) {
          folderPromises.push(
            traverseDirectory(entry as IFileSystemDirectoryEntry)
          );
        } else {
          filePromises.push(getFile(entry as IFileSystemFileEntry));
        }
      }
    } else {
      dataTransferFiles.push(new PackagedFile(listItem.getAsFile()));
    }
  });
  if (folderPromises.length) {
    return Promise.all(folderPromises).then(fileEntries => {
      const flattenedEntries = flatten(fileEntries);
      // collect async promises to convert each fileEntry into a File object
      flattenedEntries.forEach((fileEntry: any) => {
        filePromises.push(getFile(fileEntry));
      });
      return handleFilePromises(filePromises, dataTransferFiles);
    });
  } else if (filePromises.length) {
    return handleFilePromises(filePromises, dataTransferFiles);
  }
  return Promise.resolve(dataTransferFiles);
}

function flatten<T>(array: T[]): T[] {
  return array.reduce(
    (prev: T[], current: T | T[]) =>
      prev.concat(Array.isArray(current) ? flatten(current) : current),
    []
  );
}

/**
 * This function should be called from both the onDrop event from your drag/drop
 * dropzone as well as from the HTML5 file selector input field onChange event
 * handler.  Pass the event object from the triggered event into this function.
 * Supports mix of files and folders dropped via drag/drop.
 *
 * Returns: an array of File objects, that includes all files within folders
 *   and subfolders of the dropped/selected items.
 */
export function getDroppedOrSelectedFiles(
  event: React.DragEvent<HTMLDivElement>
): Promise<PackagedFile[]> {
  const dataTransfer = event.dataTransfer;

  // If the event is a data transfer, handle it this way
  if (dataTransfer && dataTransfer.items) {
    return getDataTransferFiles(dataTransfer);
  }

  // The event does not have a DataTransfer, we need to handle it
  // in a more basic way
  const files = [];
  const dragDropFileList = dataTransfer && dataTransfer.files;
  const inputFieldFileList = event.target && (event.target as any).files;
  const fileList = dragDropFileList || inputFieldFileList || [];

  for (const file of fileList) {
    files.push(new PackagedFile(file));
  }

  return Promise.resolve(files);
}

interface IFileSystemEntry {
  filesystem: IFileSystem;
  fullPath: string;
  isDirectory: boolean;
  isFile: boolean;
  name: string;
}

interface IFileSystem {}

interface IFileSystemDirectoryEntry extends IFileSystemEntry {
  createReader(): IFileSystemDirectoryReader;
  getDirectory(): IFileSystemDirectoryEntry;
  getFile(): IFileSystemFileEntry;
}

interface IFileSystemFileEntry extends IFileSystemEntry {
  file(
    successCallback: (file: File) => void,
    errorCallback?: (error: IFileError) => void
  ): void;
}

interface IFileSystemDirectoryReader {
  readEntries(
    successCallback: (entries: IFileSystemEntry[]) => void,
    errorCallback?: () => void
  ): IFileSystemEntry[];
}

interface IFileError {}
