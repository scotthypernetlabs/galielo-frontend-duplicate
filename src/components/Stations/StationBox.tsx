import { Box, Button, Grid, Typography } from "@material-ui/core";
import { Dispatch } from "redux";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  IReceiveSelectedStation,
  receiveSelectedStation
} from "../../actions/stationActions";
import { IStore } from "../../business/objects/store";
import { MyContext } from "../../MyContext";
import { PackagedFile } from "../../business/objects/packagedFile";
import { Station } from "../../business/objects/station";
import { User } from "../../business/objects/user";
import { connect } from "react-redux";
import { context } from "../../context";
import {
  faChalkboard,
  faDatabase,
  faUser
} from "@fortawesome/free-solid-svg-icons";
import { getDroppedOrSelectedFiles } from "./fileSelector";
import { linkYellow } from "../theme";
import ProgressBar from "../ProgressBar";
import React from "react";

const fileUploadTextDefault = "Browse or drop directory";

type Props = {
  pending: boolean;
  station: Station;
  receiveSelectedStation: (station: Station) => IReceiveSelectedStation;
  currentUser: User;
  history: any;
};

type State = {
  dragOver: boolean;
  disabled: boolean;
  fileUploadText: string;
  fileUpload: boolean;
  hover: boolean;
  identity: string;
};

class StationBox extends React.Component<Props, State> {
  context!: MyContext;
  constructor(props: Props) {
    super(props);
    this.state = {
      dragOver: false,
      disabled: false,
      fileUploadText: fileUploadTextDefault,
      fileUpload: false,
      hover: false,
      identity: 'Station Box'
    };
    this.handleDragOver = this.handleDragOver.bind(this);
    this.handleDragLeave = this.handleDragLeave.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
    this.stationDetails = this.stationDetails.bind(this);
    this.handleRunJobClick = this.handleRunJobClick.bind(this);
    this.handleMouseOver = this.handleMouseOver.bind(this);
    this.handleMouseOut = this.handleMouseOut.bind(this);
  }
  componentDidMount(){
    this.context.uploadQueue.bindComponent(this, this.state.identity);
  }
  componentWillUnmount(){
    this.context.uploadQueue.removeComponent(this.state.identity);
  }
  handleMouseOver(e: React.MouseEvent<HTMLDivElement, MouseEvent>) {
    e.preventDefault();
    e.stopPropagation();
    this.setState({
      hover: true
    });
  }

  handleMouseOut(e: React.MouseEvent<HTMLDivElement, MouseEvent>) {
    e.preventDefault();
    e.stopPropagation();
    this.setState({
      hover: false
    });
  }

  handleOpenStation(station: Station) {
    return (e: any) => {
      this.props.history.push(`/stations/${station.id}`);
      this.props.receiveSelectedStation(station);
    };
  }

  handleDragOver(e: React.MouseEvent<HTMLDivElement, MouseEvent>) {
    e.preventDefault();
    e.stopPropagation();
    const { disabled } = this.state;
    if (disabled) return;
    this.setState({
      fileUploadText: "Drop to send a directory",
      dragOver: true
    });
  }

  handleDragLeave(e: React.MouseEvent<HTMLDivElement, MouseEvent>) {
    e.preventDefault();
    e.stopPropagation();
    const { disabled } = this.state;
    if (disabled) return;
    this.setState({
      fileUploadText: fileUploadTextDefault,
      dragOver: false
    });
  }

  async handleDrop(e: React.DragEvent<HTMLDivElement>, station: Station) {
    e.preventDefault();
    e.stopPropagation();
    const { disabled } = this.state;
    if (disabled) return;
    this.setState({
      disabled: true,
      fileUploadText: "Uploading your file.....",
      fileUpload: true
    });
    const directoryName = e.dataTransfer.files[0].name;
    let files = await getDroppedOrSelectedFiles(e);
    files = files.map((file: PackagedFile) => {
      const path = file.fullPath.replace(`${directoryName}/`, "");
      return Object.assign({}, file, { fullPath: path.slice(1) });
    });
    let sendJobFunction = async() => {
      const jobUploaded = await this.context.jobService.sendStationJob(
        station.id,
        files,
        directoryName
      );
      this.setState({
        fileUploadText: "Queued...",
        disabled: false,
        fileUpload: false
      });
    }
    this.context.uploadQueue.addToQueue(sendJobFunction);
    this.context.uploadQueue.startQueue();
  }

  stationDetails(station: Station) {
    if (this.state.fileUpload || this.state.dragOver) {
      return (
        <Grid item xs={12}>
          <h5>{this.state.fileUploadText}</h5>
        </Grid>
      );
    }
    return (
      <>
        <Grid item={true} xs={4}>
          <FontAwesomeIcon
            icon={faChalkboard}
            style={{ color: "black", float: "left", marginRight: 5 }}
          />
          <Typography variant="h5">{station.machines.length}</Typography>
        </Grid>
        <Grid item={true} xs={4}>
          <FontAwesomeIcon
            icon={faUser}
            style={{ color: "black", float: "left", marginRight: 5 }}
          />
          <Typography variant="h5">{station.members.length}</Typography>
        </Grid>
        <Grid item={true} xs={4}>
          <FontAwesomeIcon
            icon={faDatabase}
            style={{ color: "black", float: "left", marginRight: 5 }}
          />
          <Typography variant="h5">
            {Object.keys(station.volumes).length}
          </Typography>
        </Grid>
      </>
    );
  }

  handleRunJobClick(e: React.MouseEvent) {
    e.preventDefault();
    e.stopPropagation();
    const { disabled } = this.state;
    const { station } = this.props;
    if (disabled) return;
    const inputElement = document.createElement("input");
    inputElement.type = "file";
    // @ts-ignore
    inputElement.webkitdirectory = true;
    inputElement.addEventListener("change", async file => {
      this.setState({
        fileUploadText: "Uploading your file.....",
        disabled: true
      });
      const firstFile = inputElement.files[0];
      // @ts-ignore
      const fullPath = firstFile.webkitRelativePath;
      const directoryName = fullPath.slice(
        0,
        fullPath.indexOf(`/${firstFile.name}`)
      );
      const files = Array.from(inputElement.files);
      const formattedFiles = files.map(file => {
        return Object.assign(
          {},
          {
            fileObject: file,
            // @ts-ignore
            fullPath: file.webkitRelativePath.replace(`${directoryName}/`, "")
          }
        );
      });
      let sendJobFunction = async() => {
        const jobUploaded = await this.context.jobService.sendStationJob(
          station.id,
          formattedFiles,
          directoryName
        );
        this.setState({
          fileUploadText: "Queued...",
          disabled: false
        });
      }
      this.context.uploadQueue.addToQueue(sendJobFunction);
      this.context.uploadQueue.startQueue();
    });
    inputElement.dispatchEvent(new MouseEvent("click"));
  }

  stationHoverView(station: Station) {
    let className = "station-hover-grid";
    if (!this.state.hover) {
      className += " hidden";
    }

    return (
      <Grid container className={className}>
        <Grid className="station-hover-button-container">
          <Button
            size="small"
            variant="contained"
            color="primary"
            onClick={this.handleOpenStation(station)}
          >
            View Station
          </Button>
          <Button
            size="small"
            variant="contained"
            color="primary"
            onClick={this.handleRunJobClick}
          >
            Run Job
          </Button>
        </Grid>
      </Grid>
    );
  }

  render() {
    const { station, pending } = this.props;
    return (
      <div
        onClick={this.handleOpenStation(station)}
        key={station.id}
        onDragOver={this.handleDragOver}
        onDrop={e => this.handleDrop(e, station)}
      >
        <Box
          onMouseEnter={this.handleMouseOver}
          onMouseLeave={this.handleMouseOut}
          border={1}
          borderColor="#cccccc"
          p={3}
          m={1}
          minWidth="250px"
          maxWidth="250px"
          minHeight="120px"
          maxHeight="120px"
          bgcolor="rgb(255, 255, 255, 0.5)"
          className="station-box"
        >
          <Grid container>
            <Grid item xs={12}>
              {pending ? (
                <Typography
                  gutterBottom
                  variant="h3"
                  style={{ color: linkYellow.main }}
                >
                  {station.name}
                </Typography>
              ) : (
                <Typography gutterBottom variant="h3" color="primary">
                  {station.name}
                </Typography>
              )}
            </Grid>
            {!pending && this.stationHoverView(station)}
            <Grid item xs={4}>
              <FontAwesomeIcon
                icon={faChalkboard}
                style={{ color: "black", float: "left", marginRight: 5 }}
              />
              <Typography variant="h5">{station.machines.length}</Typography>
            </Grid>
            <Grid item xs={4}>
              <FontAwesomeIcon
                icon={faUser}
                style={{ color: "black", float: "left", marginRight: 5 }}
              />
              <Typography variant="h5">{station.members.length}</Typography>
            </Grid>
            <Grid item xs={4}>
              <FontAwesomeIcon
                icon={faDatabase}
                style={{ color: "black", float: "left", marginRight: 5 }}
              />
              <Typography variant="h5">
                {Object.keys(station.volumes).length}
              </Typography>
            </Grid>
            <Grid item xs={12}>
              <ProgressBar type={"station"} id={station.id} />
            </Grid>
          </Grid>
        </Box>
      </div>
    );
  }
}

StationBox.contextType = context;

const mapStatetoProps = (state: IStore) => ({
  currentUser: state.users.currentUser
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  receiveSelectedStation: (station: Station) =>
    dispatch(receiveSelectedStation(station))
});

export default connect(mapStatetoProps, mapDispatchToProps)(StationBox);
