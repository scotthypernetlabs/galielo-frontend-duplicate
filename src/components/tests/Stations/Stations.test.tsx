import { Dictionary } from "../../../business/objects/dictionary";
import { Route } from "react-router-dom";
import { Station } from "../../../business/objects/station";
import { User } from "../../../business/objects/user";
import { createLocation, createMemoryHistory } from "history";
import { match } from "react-router";
import { shallow } from "enzyme";
import React from "react";
import Stations from "../../Stations/Stations/Stations";

describe("Stations View", () => {
  const stations: Station[] = [
    new Station(
      "stationid1",
      ["1"],
      ["1"],
      ["1", "2", "3"],
      "a",
      "description",
      ["mid1"],
      [],
      [],
      [],
      "2020-03-26T17:45:48+0000",
      "2020-03-18T16:45:48+0000"
    ),
    new Station(
      "stationid2",
      ["2"],
      ["2"],
      ["2", "1"],
      "b",
      "description",
      ["mid1", "mid2", "mid3"],
      [],
      [],
      [],
      "2020-03-25T17:45:48+0000",
      "timestamp2"
    ),
    new Station(
      "stationid1",
      ["3"],
      ["3"],
      ["3"],
      "c",
      "description",
      ["mid1"],
      [],
      [],
      [],
      "2020-03-18T17:45:48+0000",
      "timestamp2"
    )
  ];

  const stationDictionary: Dictionary<Station> = {};
  stations.forEach((station: Station) => {
    stationDictionary[station.id] = station;
  });

  const history = createMemoryHistory();
  const path = `/route/:id`;

  const match: match<{ id: string }> = {
    isExact: false,
    path,
    url: path.replace(":id", "1"),
    params: { id: "1" }
  };

  const location = createLocation(match.url);
  const componentProps = {
    slice: false,
    numberOfStations: stations.length,
    stations: stationDictionary,
    currentUser: User,
    openCreateStation: jest.fn(),
    receiveSelectedStation: jest.fn(),
    history,
    match,
    location
  };

  const stationsComponent = shallow(
    <Route>
      <Stations {...componentProps} />
    </Route>
  );

  it("matches snapshot", () => {
    // Snapshot demo
    expect(stationsComponent).toMatchSnapshot();
  });
});
