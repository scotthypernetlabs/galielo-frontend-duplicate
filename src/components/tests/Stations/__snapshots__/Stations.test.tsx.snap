// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stations View matches snapshot 1`] = `ShallowWrapper {}`;
